orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java

```diff
@@ -556,46 +559,55 @@ public void installFreeIpa(GatewayConfig primaryGateway, List<GatewayConfig> all
-            Set<String> existingFreeIpaReplaceHostnames, Set<Node> allNodes, ExitCriteriaModel exitCriteriaModel) throws Exception {
-        freeIpaMasterHostname = new HashSet<>(freeIpaMasterHostname);
-        if (!freeIpaMasterHostname.isEmpty()) {
-            LOGGER.debug("Existing primary FreeIPA: {}", freeIpaMasterHostname);
-        } else if (existingFreeIpaReplaceHostnames.isEmpty()) {
-            freeIpaMasterHostname.add(primaryGateway.getHostname());
-            LOGGER.debug("Initial primary FreeIPA: {}", freeIpaMasterHostname);
+            Set<String> existingFreeIpaReplicaHostnames, Set<Node> allNodes, ExitCriteriaModel exitCriteriaModel) throws Exception {
+        Set<String> freeIpaMasterHostnames = new HashSet<>(freeIpaMasterHostname);
+        if (!freeIpaMasterHostnames.isEmpty()) {
+            LOGGER.debug("Existing primary FreeIPA: {}", freeIpaMasterHostnames);
+        } else if (existingFreeIpaReplicaHostnames.isEmpty()) {
+            freeIpaMasterHostnames.add(primaryGateway.getHostname());
+            LOGGER.debug("Initial primary FreeIPA: {}", freeIpaMasterHostnames);
             saltCommandRunner.runModifyGrainCommand(sc,
-                    new GrainAddRunner(freeIpaMasterHostname, allNodes, FREEIPA_MASTER_ROLE), exitCriteriaModel, exitCriteria);
+                    new GrainAddRunner(freeIpaMasterHostnames, allNodes, FREEIPA_MASTER_ROLE), exitCriteriaModel, exitCriteria);
         } else {
-            freeIpaMasterHostname.add(unassignedHostnames.stream().findFirst()
-                    .orElseThrow(() ->
-                            new NotFoundException("A primary FreeIPA instance is requried and there are no unassigned roles to assign as a primary")));
-            LOGGER.debug("Replacement primary FreeIPA: {}", freeIpaMasterHostname);
+            freeIpaMasterHostnames.add(unassignedHostnames.stream().findFirst()
+                    .orElseGet(() -> choosePrimaryGatewayAsMasterReplacement(primaryGateway, existingFreeIpaReplicaHostnames)
+                            .orElseThrow(() ->
+                                    new NotFoundException("A primary FreeIPA instance is required and there are no unassigned roles to assign as a primary"))));
+            LOGGER.debug("Replacement primary FreeIPA: {}", freeIpaMasterHostnames);
             saltCommandRunner.runModifyGrainCommand(sc,
-                    new GrainAddRunner(freeIpaMasterHostname, allNodes, FREEIPA_MASTER_REPLACEMENT_ROLE), exitCriteriaModel, exitCriteria);
+                    new GrainAddRunner(freeIpaMasterHostnames, allNodes, FREEIPA_MASTER_REPLACEMENT_ROLE), exitCriteriaModel, exitCriteria);
+            saltCommandRunner.runModifyGrainCommand(sc, new GrainRemoveRunner(freeIpaMasterHostnames, allNodes, FREEIPA_REPLICA_ROLE), exitCriteriaModel,
+                    exitCriteria);
```

Comment:
I would extract that to a method with a good name (like promoteAReplicaNodeToMaster or something else).